src/frontend.rs

```diff
@@ -0,0 +1,73 @@
+use std::path::Path;
+
+use crate::cache::ModuleCache;
+use crate::lexer::Lexer;
+use crate::nameresolution::NameResolver;
+use crate::parser;
+use crate::types;
+use crate::util;
+
+#[derive(PartialEq)]
+pub enum FrontendPhase {
+    Lex,
+    Parse,
+    TypeCheck,
+}
+
+pub enum FrontendResult {
+    Done,
+    ContinueCompilation,
+    Errors,
+}
+
+pub fn check<'a>(
+    filename: &'a Path, main_file_contents: String, cache: &mut ModuleCache<'a>, phase: FrontendPhase, show_time: bool,
```

Comment:
```suggestion
    filename: &'a Path, main_file_contents: String, cache: &mut ModuleCache<'a>, stop_after_phase: FrontendPhase, show_time: bool,
```
`phase` sounded a bit too general to me